src/Modules/Visualization/ShowField.cc

```diff
@@ -430,6 +430,96 @@ void GeometryBuilder::renderFaces(
 
 
 
+inline static void writeAtributeToVBO(const Point& point, spire::VarBuffer* vboBuffer)
+{
+  vboBuffer->write(static_cast<float>(point.x()));
+  vboBuffer->write(static_cast<float>(point.y()));
+  vboBuffer->write(static_cast<float>(point.z()));
+}
+
+inline static void writeAtributeToVBO(const Vector& vector, spire::VarBuffer* vboBuffer)
+{
+  vboBuffer->write(static_cast<float>(vector.x()));
+  vboBuffer->write(static_cast<float>(vector.y()));
+  vboBuffer->write(static_cast<float>(vector.z()));
+}
+
+inline static void writeAtributeToVBO(const ColorRGB& color, spire::VarBuffer* vboBuffer)
+{
+  vboBuffer->write(static_cast<float>(color.r()));
+  vboBuffer->write(static_cast<float>(color.g()));
+  vboBuffer->write(static_cast<float>(color.b()));
+  vboBuffer->write(static_cast<float>(1.0f));
+}
+
+inline static void writeAtributeToVBO(const glm::vec2& coords, spire::VarBuffer* vboBuffer)
+{
+  vboBuffer->write(static_cast<float>(coords.x));
+  vboBuffer->write(static_cast<float>(coords.y));
+}
+
+inline static void writeIndexToIBO(const uint32_t index, spire::VarBuffer* iboBuffer)
+{
+  iboBuffer->write(index);
+}
+
+template<typename ... Params>
+static void writeTri(
+  spire::VarBuffer* vboBuffer,
+  spire::VarBuffer* iboBuffer,
+  uint32_t& iboIndex,
+  const Params& ... params)
```

Comment:
I applaud the template attempt but there's an easier+clearer way to handle the optional arguments.

The points vector is always passed in. So make that a normal parameter first.

Then the question is how to add 0,1, or 2 of normals and textureCoords. Just make those boost::optional<vector>s, and pass in `{}` for the cases where you don't need them. This function then looks like
```
static void writeTri(
  spire::VarBuffer* vboBuffer,
  spire::VarBuffer* iboBuffer,
  uint32_t& iboIndex,
const std::vector<Point>& points,
 boost::optional<std::vector<Vector>> normals,
  boost::optional<std::vector<glm::vec2>> textureCoords)
{
for 0->2
   write point
   if (normals)
      write normal
   if (textureCoords)
      write Texture
```
with no template. The switch looks like
```
        case JUST_POINTS: writeTri(vboBuffer, iboBuffer, iboIndex, points, {}, {}); break;
        case WITH_COLOR: writeTri(vboBuffer, iboBuffer, iboIndex, points, {}, textureCoords); break;
        case WITH_NORMALS: writeTri(vboBuffer, iboBuffer, iboIndex, points, normals, {}); break;
        case WITH_COLOR_AND_NORMALS: writeTri(vboBuffer, iboBuffer, iboIndex, points, normals, textureCoords); break;
```

Easier to read, yes? No need for "janky" voided initializer_list expanded parameter pack.

Reply:
Just make sure the vectors aren't copied when passing them as boost::optional parameters. Doc here: https://www.boost.org/doc/libs/1_63_0/libs/optional/doc/html/boost_optional/tutorial/performance_considerations.html

Reading that, you can use optional references:
`boost::optional<const std::vector<Vector>&> normals,`
etc

Reply:
Not quite. This would defeat the whole point of me writing it the way I did. I wanted each one to be its own function so I could move the conditional statements out of the write functions (to replace with the switch statement). This was done for performance reasons since this is the tightest loop of the face generation code. And while the void initializer is janky it was again done for performance reasons. Writing a recursive template destroyed performance so I used an initializer list instead (hopefully there will be a cleaner way of doing this in the future but currently there is not).